rfd/0215-mwi-terraform-provider.md

```diff
@@ -0,0 +1,357 @@
+choose their own version, other teams are leveraging OpenTofu or platforms
+which currently use older versions of Terraform.
+
+Therefore, we should support both data sources and ephemeral resources at least
+until support is more common-place. This will enable customers with more recent
+versions of Terraform to use the more secure option. In 6-12 months, we can 
+revisit this decision and consider deprecating support for data sources.
+
+### Which provider to use
+
+One key decision is where to build this functionality. Largely, we have three
+choices:
+
+- Build this functionality within the existing Teleport Terraform provider.
+  - Positives:
+    - No additional build/release pipelines to maintain.
+    - Customers leveraging this functionality and managing a Teleport cluster 
+      within the same plan/apply only need to configure a single provider.
+  - Negatives:
+    - Existing Teleport terraform provider has a significant degree of tech
+      debt, and due to the different nature of this functionality, we're likely
+      to run into unforeseen edge-cases within the core implementation of the 
+      Terraform provider.
+    - Existing Terraform provider consists largely of generated code, whereas
+      this functionality will be largely hand-written. Mixing generated and 
+      hand-written code can make enforcing standards more difficult.
+    - Existing Terraform provider uses a very old version of the Terraform SDK
+      which makes compatability with other tools like Pulumi more challenging.
+    - Existing Terraform provider has a large number of configuration
+      parameters which would not be compatible with this functionality, this
+      poses a risk of creating a poor configuration UX.
+- Build all variants of this functionality into a new Terraform provider
+  - Positives:
+    - No inherited tech debt from existing provider & can leverage latest
+      versions of the Terraform SDK.
+    - Separation of hand-written and generated code.
+  - Negatives:
+    - Additional build and release infrastructure to maintain.
+    - Additional codebase to maintain.
+- Build each variant of this functionality as its own Terraform provider
+  - Positives:
+    - No inherited tech debt from existing provider & can leverage latest
+      versions of the Terraform SDK.
+    - Separation of hand-written and generated code.
+    - Separation of variants would allow for smaller binary sizes.
+  - Negatives:
+    - Much more build and release infrastructure to maintain. This would need 
+      to be handled each time we added a new variant.
+
+Overall, the most appropriate approach seems to be introducing a single new
+Terraform provider for this functionality.
+
+## UX
+
+Before using the data sources or ephemeral resources, the user must configure
+the provider itself. To do so, they will provide the details required to
+connect and authenticate to the Teleport cluster.
+
+Configuring the MWI provider:
+
+```hcl
+provider "mwi" {
```

Comment:
I'm a little unkeen on the name "mwi" so if someone has a nicer, but still concise-ish name, that'd be great. Conciseness is key since the provider name is prefixed to resource/data source kind names.

Reply:
Naming things, caching and off-by-one problems...

I would like `teleport_bot` except it's the identical name of a resource from our existing provider. Are people gonna configure Teleport in the same Terraform as use this? Maybe not but it feels bad to potentially search `teleport_bot` and get two kinds of things in a codebase.

Extending that, I don't think we'd end up a with another Terraform resource for IS or IG that would be called exactly `teleport_identity`, so that one feels the best to me? Especially because `provider "teleport_identity` {` reads nicely; the identity is what you're providing. Maybe even `teleport_identities`?

Reply:
I actually have a feeling I missed that Terraform requires that the provider name is a single word - "must contain only letters, digits, and dashes, and may not use leading or trailing dashes". Looking thru the public registry, I can't see any that use a dash or underscore.